File: OG/P3_Token.java
import java.util.*;
class token
{
static boolean isValid(String str, int n)
{

	// If first character is invalid
	if (!((str.charAt(0) >= 'a' && str.charAt(0) <= 'z')
		|| (str.charAt(0)>= 'A' && str.charAt(0) <= 'Z')))
		return false;

	// Traverse the string for the rest of the characters
	for (int i = 1; i < str.length(); i++)
	{
		if (!((str.charAt(i) >= 'a' && str.charAt(i) <= 'z')
			|| (str.charAt(i) >= 'A' && str.charAt(i) <= 'Z')
			|| (str.charAt(i) >= '0' && str.charAt(i) <= '9')
			|| str.charAt(i) == '_'))
			return false;
	}

	// String is a valid identifier
	return true;
}

public static boolean opr(String str)
{
	String[] a={"+","-","/","*","<=",">=","<",">","==","!="};
	int i;
	boolean flag=false;
	for(i=0;i<a.length;i++)
	{
		boolean c=str.equals(a[i]);
		if(c)
		{
		flag=true;
		break;
		}
		else	
		flag=false;	
	}
return flag;
}


public static boolean punc(String str)
{
	String[] a={".",",",";","(",")","{","}","[","]"};
	int i;
	boolean flag=false;
	for(i=0;i<a.length;i++)
	{
		boolean c=str.equals(a[i]);
		if(c)
		{
			flag=true;
			break;
		}
		else	
		flag=false;	
	}
return flag;
}

public static boolean res(String str)
{
	String[] a={"printf","scanf","if","else","break"};
	int i;
	boolean flag=false;
	for(i=0;i<a.length;i++)
	{
		boolean c=str.equals(a[i]);
		if(c)
		{
		flag=true;
		break;
		}
		else	
		flag=false;	
	}
return flag;
}

public static boolean pcon(String str)
{
	boolean flag=true;
	for (int i = 0; i < str.length(); i++) 
    	{
		if (!(str.charAt(i)>='0' && str.charAt(i)<='9'))
        	{
		flag=false;
		}
	
	}
return flag;
}

public static void main(String args[])
{
	System.out.println("Enter a string");
	Scanner c=new Scanner(System.in);
	String str = c.nextLine();
	int n = str.length();

	if (isValid(str, n)||res(str)||punc(str)||opr(str)||pcon(str))
		System.out.println(str+" Is a Token");
	else
		System.out.println(str+" is Not a Token");
}
}

File: Py/P3_Token.py
#Design a lexical analyzer to recognize the token defined by the given program.
#Saail Chavan, Saail007

def is_valid_identifier(s):
    # Check if the first character is a letter
    if not (('a' <= s[0] <= 'z') or ('A' <= s[0] <= 'Z')):
        return False

    # Traverse the string for the rest of the characters
    for char in s[1:]:
        if not (('a' <= char <= 'z') or ('A' <= char <= 'Z') or ('0' <= char <= '9') or char == '_'):
            return False

    # String is a valid identifier
    return True

def is_operator(s):
    operators = ["+", "-", "/", "*", "<=", ">=", "<", ">", "==", "!="]
    return s in operators

def is_punctuation(s):
    punctuations = [".", ",", ";", "(", ")", "{", "}", "[", "]"]
    return s in punctuations

def is_reserved_keyword(s):
    reserved_keywords = ["printf", "scanf", "if", "else", "break"]
    return s in reserved_keywords

def is_constant(s):
    # Check if all characters are digits
    for char in s:
        if not ('0' <= char <= '9'):
            return False
    return True

def main():
    str_input = input("Enter a string: ")
    # Check if the string is a valid identifier, reserved keyword, operator, punctuation, or constant
    if (is_valid_identifier(str_input) or is_reserved_keyword(str_input) or
            is_punctuation(str_input) or is_operator(str_input) or is_constant(str_input)):
        print(f"{str_input} is a Token")
    else:
        print(f"{str_input} is Not a Token")

if __name__ == "__main__":
    main()
